ForumProject/users/views.py

```diff
@@ -60,6 +66,76 @@ def post(self, request, *args, **kwargs):
+            user = request.user
+            if UserRoleCompany.objects.filter(user=user).exists():
+                user_role = UserRoleCompany.objects.get(user=user)
+                user_role.role = serializer.validated_data['role']
+                user_role.company_id = 0
+                user_role.save()
+            else:
+                UserRoleCompany.objects.create(user=user, role=serializer.validated_data['role'])
+            return Response({'success': 'Role has been successfully updated.'}, status=status.HTTP_200_OK)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class CompanySelectionView(APIView):
+    permission_classes = [IsRole]
+
+    def post(self, request):
+        serializer = CompanySerializer(data=request.data)
+        if serializer.is_valid():
+            user = request.user
+            company = serializer.validated_data['company_id']
+
+            user_role_company = UserRoleCompany.objects.get(user=user)
+
+            if user_role_company.role == 'investor':
+                user_investors = UserInvestor.objects.filter(customuser=self.request.user).values('investor')
+                if {'investor': company} in user_investors:
+                    user_role_company.company_id = company
+                    user_role_company.save()
+                    return Response({'success': 'The investor company was successfully selected'}, status=status.HTTP_200_OK)
+
+            elif user_role_company.role == 'startup':
+                user_startups = UserStartup.objects.filter(customuser=self.request.user).values('startup')
+                if {'startup': company} in user_startups:
+                    user_role_company.company_id = company
+                    user_role_company.save()
+                    return Response({'success': 'The startup company was successfully selected'}, status=status.HTTP_200_OK)
+            return Response({'error': 'You must select only your company'}, status=status.HTTP_400_BAD_REQUEST)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class UserCompanyView(generics.ListAPIView):
+    permission_classes = []  # write permission only for users with role
+
+    def get_serializer_class(self):
+        user = UserRoleCompany.objects.get(user=self.request.user)
+        user_role = user.role
+
+        if user_role == 'startup':
+            return StartupSerializer
+        elif user_role == 'investor':
+            return InvestorSerializer
+
+    def get_queryset(self):
+        user = UserRoleCompany.objects.get(user=self.request.user)
+        user_role = user.role
+
+        if user_role == 'startup':
+            user_startups = UserStartup.objects.filter(customuser=self.request.user).values('startup')
+            return Startup.objects.filter(id__in=user_startups)
+        elif user_role == 'investor':
+            user_investors = UserInvestor.objects.filter(customuser=self.request.user).values('investor')
```

Comment:
When building the queryset, using `values_list('startup', flat=True)` would be more appropriate and efficient than `values('startup')`. The same for `values('investor')`

Reply:
Thanks. Good idea.